text/0000-array-private-classes.md

```diff
@@ -0,0 +1,366 @@
+    else
+      error
+    end
+```
+
+to
+
+```pony
+  fun values(): CollectionIterator[Xml2node]^ ? =>
+    if (allocated) then
+      nodearray.values()
+    else
+      error
+    end
+```
+
+In this sample, the developer was not really concerned by the type of the iterator
+but that the `values` function must return an `CollectionIterator` over `Xml2node`.
+The new version makes the code simpler to understand.
+
+This change in `array.pony` and other collections will break such code but it
+can be easily adapted to use the new API. And it will make the standard library
+easier to learn by reducing the number of public types.
+
+# How We Test This
+
+Pony tests must continue to pass. No additional tests are need as after review
+the existing coverage in Pony standard library tests is sufficient
+
+# Drawbacks
+
+Will break any existing code that uses any of the classes that are currently
+public and will be made private by this RFC.
+
+# Alternatives
+
+1. Stay as is. Continue the
+[discussion on Zulip](https://ponylang.zulipchat.com/#narrow/stream/189959-RFCs/topic/Make.20Array.20iterators.20private).
+
+2. Update the existing concrete classes to include `rewind` where needed.
+
+3. Instead of defining a new type of iterator with the `CollectionIterator` interface,
+we can consider only the rewindable part of it in an `Rewindable` interface:
+
+```pony
+interface Rewindable[A]
+  fun ref rewind(): A^
+    """
+    Rewind the type `A`.
+    """
+```
+
+Then one can create a rewindable iterator by combining these two interfaces into
+a intersection type:
+
+```pony
+type CollectionIterator[A] is (Iterator[A] & Rewindable[Iterator[A]])
+```
+
+Perhaps, a more general interface name instead of `Rewindable` would be `Resetable`
+to define the traits of a type that can be reset to its initial state, as in the
+case of iterators `reset` == `rewind`.
+
+The rewindable type can be used with other types than `Iterator`, like a data
+structure that would implement a rewindable property. This alternative was
+[put aside](https://github.com/ponylang/rfcs/pull/193#discussion_r780793165) to
+prevent name colisions in `builtin` with user-named types.
+
+## Comparison of alternatives
+
+1. Do nothing: return public concrete classes.
+2. Update the existing return classes and add `rewind`.
+3. Add a `CollectionIterator` interface for collections.
+4. Create intersection type by defining a `Rewindable` interface and combine
+with `Iterator`.
+5. Change `Iterator` to add `rewind`.
+
+| Justification        /      Alternative # |  1  |  2  |  3  |  4  |  5  |
+|:------------------------------------------|:---:|:---:|:---:|:---:|:---:|
+| a. Simpler stdlib API                     | --- | --- |  +  | +++ | +++ |
+| b. Simpler function signatures            | --- | --- |  +  | +++ | +++ |
+| c. Stay compatible with existing stdlib   | +++ | +++ |  +  | +++ |  +  |
+| d. Evolutive                              | --- |  +  | +++ | +++ |  +  |
```

Comment:
Changes for the sake of changes is not evolution -- there must be positive or negative selection which lead microevolution into macroevolution. Put another way, if all the suggested changes have negative effects than it is still "evolution" to do nothing as we do not bring those negative effects forward with us.

Side note: I am a career "full stack biologist" who has done work ranging from cancer to conservation to COVID-19 -- Pony is the project that I volunteer my time with when **not** working on the "evolving nature" of biological problems.